src/libraries/System.Security.Cryptography/docs/X509CertificateLoader.SecurityDesign.md

```diff
@@ -0,0 +1,335 @@
+While most of these formats can be processed in _O(n)_ time, based on the length of the input, the PKCS#12 PFX format contains work-factor values within it, meaning that no running-time assumptions can be made for a context-free call to `new X509Certificate2(bytes)`.
+
+Rather than single PKCS#12 PFX out as the one thing to remove, the design of X509CertificateLoader is to limit each method group to the support of a single file format.
+
+## File Encodings
+
+Both the X.509 file format and the PKCS#7 SignedData format support a binary (DER/BER) and a textual (PEM) representation.
+None of the rest of the formats supported by the now-legacy constructors support a textual representation.
+
+### Encoding-Sniffing
+
+Any method group on `X509CertificateLoader` which is processing a format that supports both a PEM and a BER/DER encoding will sniff the encoding, rather than force users to be cognizant of the difference (if requested by users, we may in the future allow single-encoding variants of these methods).
+
+If the first byte of the provided data (either file or in-memory representation) is `0x30` then the data will be interpreted in the binary encoding, otherwise it will be interpreted in the PEM encoding.
+
+Callers implementing a protocol or file format where only one of the two encodings is acceptable will need to do their own validation.
+
+### Binary Encoding
+
+The now-legacy constructors and the `SignedCms.Decode` method all behave by reading data starting at the first byte, and stop at the last _semantic_ byte (as determined by the payload contents).
+While strictness is a virtue, so is compatibility, and it is in the interest of compatibility (not rejecting inputs that are identified as the correct format by `X509Certificate2.GetCertContentType`) that `X509CertificateLoader` functions in the same manner.
+
+That is to say, supposing the byte sequence `30 02 05 00` were a valid representation of a format supported by `X509CertificateLoader` (which it is not), the loader will have identicial behavior for that input and for `30 02 05 00 BE EF`, silently discarding the two trailing bytes.
+
+Callers implementing a protocol or file format where the trailing data is unacceptable will need to perform their own content-length validation.
+
+#### BER, CER, or DER
+
+Each file format uses the binary encoding most appropriate to that format.
+For example, ITU-T X.509 and IETF RFC 3280 both indicate that the `Certificate` data type must always use the ASN.1 Distinguished Encoding Rules (DER) restrictions; therefore `LoadCertificate` requires a DER encoding.
+Alternatively, IETF RFC 7292 describes PKCS#12 PFX as being in the relaxed Basic Encoding Rules (BER) form, so `LoadPkcs12` only requires BER.
+
+Callers implementing a protocol or file format where either Canonical Encoding Rules (CER) or DER are required where the format is specified only as BER will need to perform their own restricted-encoding validation.
+
+### Textual Encoding
```

Comment:
A note that PEM encoding must conform to RFC 7468 "lax" might be good. "PEM" by itself has a lot of ambiguity because some "other" PEMs permit metadata like GPGs.